### PR TITLE
ExampleMod updated with Asset guidelines

### PR DESCRIPTION
**Pull request by JavidPack**
In preparing [the Asset wiki page](https://github.com/tModLoader/tModLoader/wiki/Assets), it was discovered that ExampleMod didn't always show the recommended approach to dealing with `Asset`s. This PR updates ExampleMod to follow the guidelines in an attempt at showing good code design and efficient Asset usage.

This PR will be updated if needed based on feedback on the new approaches.

### Main.instance.LoadX
ExampleMod uses `Main.instance.LoadX` methods in various places where it is unnecessary. All modded content textures are loaded during mod loading. These have been removed

### Main.dedServ gates
No need to check `Main.dedServ` before calling `Request`. These have been removed.

### Remove code setting Assets to null
There is no need to set `Asset` fields to `null`. These have been removed
```diff
-public override void Unload() {
-	chainTexture = null;
-}
```

### Don't use ImmediateLoad unless needed
`, AssetRequestMode.ImmediateLoad` has been removed from `Request`s where it was not necessary.

### Store Extra Textures in Backing field
Backing fields added to reduce `Request` calls
```diff
+private Asset<Texture2D> rainTexture;
+public override void Load() {
+	rainTexture = Mod.Assets.Request<Texture2D>("Content/Biomes/ExampleRain");
+}
+public override Asset<Texture2D> GetRainTexture() => rainTexture;

-public override Asset<Texture2D> GetRainTexture() {
-	return ModContent.Request<Texture2D>("ExampleMod/Content/Biomes/ExampleRain");
-}
```

### Request assets during load instead of on demand 
To catch typos during mod loading, assets are loaded during loading in most cases.
```diff

+private static Asset<Texture2D> chainTexture;
+private static Asset<Texture2D> chainTextureExtra; // This texture and related code is optional and used for a unique effect

+public override void Load() {
+	chainTexture = ModContent.Request<Texture2D>(ChainTexturePath);
+	chainTextureExtra = ModContent.Request<Texture2D>(ChainTextureExtraPath);
+}

public override bool PreDraw(ref Color lightColor) {
-	Asset<Texture2D> chainTexture = ModContent.Request<Texture2D>(ChainTexturePath);
-	Asset<Texture2D> chainTextureExtra = ModContent.Request<Texture2D>(ChainTextureExtraPath); // This texture and related code is optional and used for a unique effect
	// etc
```

### CODE REVIEW

File: ExampleMod/Content/NPCs/ExampleBoneMerchant.cs [hunk 1]
```diff
@@ -24,6 +25,15 @@ namespace ExampleMod.Content.NPCs
...
 		private static Profiles.StackedNPCProfile NPCProfile;
+		private static Asset<Texture2D> shimmerGun;
+
+		public override void Load() {
+			shimmerGun = ModContent.Request<Texture2D>(Texture + "_Shimmer_Gun");
+		}
+
+		public override void Unload() {
+			shimmerGun = null;
+		}
```
Comment (Chicken-Bones) [inline ExampleMod/Content/NPCs/ExampleBoneMerchant.cs]: Not necessary, I know we don't unload ALCs properly yet, but I don't want to mislead modders into thinking they have to clean up every static field they ever make

File: ExampleMod/Content/Projectiles/ExampleAdvancedAnimatedProjectile.cs [hunk 2]
```diff
@@ -105,18 +106,18 @@ public class ExampleAdvancedAnimatedProjectile : ModProjectile
 				spriteEffects = SpriteEffects.FlipHorizontally;
 
 			// Getting texture of projectile
-			Texture2D texture = (Texture2D)ModContent.Request<Texture2D>(Texture);
+			Texture2D texture = TextureAssets.Projectile[Type].Value;
```
Comment (Chicken-Bones) [inline ExampleMod/Content/Projectiles/ExampleAdvancedAnimatedProjectile.cs:109]: Hm, it seems that modded entries in `TextureAssets.Projectile/Npc/Item` will always be loaded, where vanilla entries will be loaded on-demand via `Main.LoadProjectile/Item/NPC`.

I'm not sure what we should do about this, but it's a nuance that seems like a gotcha.

File: ExampleMod/Content/Projectiles/ExampleAdvancedFlailProjectile.cs [hunk 3]
```diff
@@ -39,6 +42,16 @@ private enum AIState
 		public ref float CollisionCounter => ref Projectile.localAI[0];
 		public ref float SpinningStateTimer => ref Projectile.localAI[1];
 
+		public override void Load() {
+			chainTexture = ModContent.Request<Texture2D>(ChainTexturePath);
+			chainTextureExtra = ModContent.Request<Texture2D>(ChainTextureExtraPath);
+		}
```
Comment (Chicken-Bones) [inline ExampleMod/Content/Projectiles/ExampleAdvancedFlailProjectile.cs:48]: Sometimes we `Request` on `Load`, sometimes on demand. We should give some advice, or pick one and stick to it

File: ExampleMod/Content/Tiles/ExampleTorch.cs [hunk 4]
```diff
@@ -69,9 +69,7 @@ public class ExampleTorch : ModTile
 			AddMapEntry(new Color(200, 200, 200), Language.GetText("ItemName.Torch"));
 
 			// Assets
-			if (!Main.dedServ) {
-				flameTexture = ModContent.Request<Texture2D>("ExampleMod/Content/Tiles/ExampleTorch_Flame");
-			}
+			flameTexture = Mod.Assets.Request<Texture2D>("Content/Tiles/ExampleTorch_Flame");
```
Comment (Chicken-Bones) [inline ExampleMod/Content/Tiles/ExampleTorch.cs]: `TextureAssets.ItemFlame` exists

File: ExampleMod/Content/Tiles/Plants/ExampleTree.cs [hunk 5]
```diff
@@ -41,12 +45,12 @@ public class ExampleTree : ModTree
 
 		// Branch Textures
 		public override Asset<Texture2D> GetBranchTextures() {
-			return ModContent.Request<Texture2D>("ExampleMod/Content/Tiles/Plants/ExampleTree_Branches");
+			return branchesTexture ??= ModContent.Request<Texture2D>("ExampleMod/Content/Tiles/Plants/ExampleTree_Branches");
```
Comment (Chicken-Bones) [inline ExampleMod/Content/Tiles/Plants/ExampleTree.cs]: Do we want to show using `Mod.Assets.Request` here? Or are we happy leaving that for 'advanced' modders and sticking to `ModContent` with full paths?

File: patches/tModLoader/Terraria/ModLoader/ModContent.cs [hunk 6]
```diff
@@ -103,6 +103,8 @@ public static bool FileExists(string name)
 	/// <summary>
 	/// Gets the asset with the specified name. Throws an Exception if the asset does not exist.
 	/// <para/>
+	/// Can be safely called on the server despite the server not loading assets, so there is no need to add checks for <see cref="Main.dedServ"/>.
+	/// <para/>
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/ModContent.cs]: Belongs in the inherited `IAssetRepository.Request` docs. Perhaps with a note that on the server assets will always be returned as if loaded via `DoNotLoad`.

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]

File: ExampleMod/Content/Tiles/Plants/ExampleTree.cs (hunk 5 above)
Comment (JavidPack) [reply ExampleMod/Content/Tiles/Plants/ExampleTree.cs]: The plant classes don't actually have a Mod field, it seems.

File: ExampleMod/Content/Tiles/ExampleTorch.cs (hunk 4 above)
Comment (JavidPack) [reply ExampleMod/Content/Tiles/ExampleTorch.cs]: This is a tile, there is a TextureAssets.Flames used for tiles, but they have a unique index and just complicate things. `Main.tileFlame` docs say not to use them.

File: ExampleMod/Content/Projectiles/ExampleAdvancedFlailProjectile.cs (hunk 3 above)
Comment (JavidPack) [reply ExampleMod/Content/Projectiles/ExampleAdvancedFlailProjectile.cs:48]: Most are on load now, I think that is best.

**Review by Chicken-Bones — APPROVED** [review]

**Pull request merged**